### ISSUE TITLE
Use .getMap() outside of Map Component 

### ISSUE THREAD

**Issue by l3git9** [visgl/react-map-gl#904]
Hi,

Working on a project that is getting quite large and I was wondering if there is a simple solution to using the map outside of the component. At the moment I am using a ref in order to use getMap() within the map component, but I have a search function that needs to flyTo the location when the user clicks on the result and I would like to be able to do so within another component. Is that possible or is everything going to need to funnel back into the Map component? I've looked everywhere for an example and have found nothing so I believe that I won't to in the manner that I want to do it. If I can't that's fine, but it would suck if there is and I just absolutely pass it up only to make it more complicated.

**Comment by mayteio** [comment]
I use context inside React to get access to the map instance:

```javascript
// MapContext.js
import React, {createContext, useState} from 'react';

export const MapContext = createContext();

export function MapProvider(props){
  const [map, setMap] = useState();
  return <MapContext.Provider value={[map, setMap]} {...props} />
}

export function useMap(){
  const context = useContext(MapContext);
  if(context === undefined) throw Error('You forgot to wrap your app with <MapProvider />');
  return context;
}

// App.js
import React from 'react';
import {MapProvider} from './MapContext';

export default function App(){
  const [, setMap] = useMap();
  return (
    <MapProvider>
      <Example />
      <ReactMapGL
        ...
        ref={ref => ref && setMap(ref)}
      />
    </MapProvider>
  )
}

// Example.js
import React from 'react';
import {useMap} from './MapContext';

export default function Example(){
  const [map] = useMap();
  const onClick = () => {
    map && map.flyTo(...);.
  }

  useEffect(() => alert('map ready!'), [map])

  return <button onClick={onClick}>Fly like a bird</button>
}
```

Note; I have switched between using `useRef` and `useState` to store the context. They both seem to have tradeoffs; [useRef is probably the intended usage according to dan](https://twitter.com/dan_abramov/status/1105930134186532864?lang=en), but this makes it difficult to listen to it actually being set, i.e. with `useEffect`. [Depending on ref.current is problematic though](https://github.com/facebook/react/issues/14387#issuecomment-503616820).

useState on the other hand will trigger an effect re-run. Mutations to the map object should not trigger a re-render because it's the same object, just mutated.

I believe the correct way is to [use a callbackRef](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node), however I am yet to implement this for `react-map-gl`.

Maybe a feature request to expose a useMap hook? 🙏

**Comment by edonadei** [comment]
Hello ! I would like to re-open this one, i'm having issue with own code base to get ref on the map to use the method flyTo. The example ahead just does not work, my code simplified

`
import ReactMapboxGl, { Layer, Feature, Marker } from "react-mapbox-gl";

const Map = ReactMapboxGl({
  accessToken: process.env.REACT_APP_MAPBOX_API
});

<Map
        // eslint-disable-next-line
        ref={ref => console.log(ref)}
        style="mapbox://styles/mapbox/streets-v9"
        containerStyle={{
          height: "93vh",
          width: "100%"
        }}
      />
`

I tried to find the method flyTo in this ref, but could not find it.

**Comment by mayteio** [comment]
This is not going to work because you’re trying to instantiate a component instead of render it. Use it as a component and pass mapboxApiAccessToken as a prop instead.

**Comment by edonadei** [comment]
Sorry for your fast answer @mayteio, just a mistake from me, i've just mistaken your library with react-mapbox-gl !!

**Comment by Pessimistress** [comment]
> Maybe a feature request to expose a useMap hook? 

Yes! All PRs welcomed.

**Comment by mayteio** [comment]
I'd be happy to @Pessimistress. What are your thoughts on API design? My solution above requires a context (I think all solutions would?) to provider the ref down to other components. That would mean introducing a `<MapRefProvider />` or something, increasing API surface. Are you opposed to this approach?

**Comment by Pessimistress** [comment]
[MapContext](https://github.com/visgl/react-map-gl/blob/850ae97d54671ea43bad11741cb05bde24caa362/docs/api-reference/map-context.md) will be available in v5.3/v6.1. You can try it with `6.1.0-beta.1`.

**Comment by emekaokoli** [comment]
Hello, my problem landed me here. please can anyone help me with this?
 https://github.com/visgl/react-map-gl/discussions/1983

the map renders but it does not update the mapref when I click to open the modal thereby making the mapref.current null which means I cannot access flyTo() or setCenter()

I will greatly appreciate all the help I can get.

thanks.

**Comment by bertearazvan** [comment]
> [MapContext](https://github.com/visgl/react-map-gl/blob/850ae97d54671ea43bad11741cb05bde24caa362/docs/api-reference/map-context.md) will be available in v5.3/v6.1. You can try it with `6.1.0-beta.1`.

It's not exported anymore unfortunately, I guess since v7?

**Comment by chris-wasowicz** [comment]
Bump, I would also like to find some good way of accessing the Map outside of the `Map` component children. As @bertearazvan mentioned, `MapContext` is no longer exported. Any other way @Pessimistress ?
